File: adapters/ruby/lib/prauga/flexdoc/host_execution_observability.rb
# frozen_string_literal: true

require "time"

module Prauga
  module FlexDoc
    # Execution evidence for the Ruby host executor.
    #
    # This mirrors the Node, Python, Go and Rust contract deliberately: the same
    # reason vocabulary, the same metric names and labels, and the same export
    # schema. An operator running a mixed fleet should read one document shape
    # regardless of which runtime served the execute route, and a collector
    # written for one runtime should not need a second parser for another.
    module HostExecutionObservability
      # Stable low-cardinality categories for a non-successful execution.
      #
      # Rejection messages interpolate request values such as origins, field
      # names and methods, so they are unbounded and cannot be used as a metric
      # label. These can.
      HOST_EXECUTION_REASONS = %w[
        marker-missing
        execution-disabled
        admission-saturated
        destination-forbidden
        redirect-forbidden
        body-malformed
        body-too-large
        unsupported-media-type
        request-invalid
        auth-unsupported
        upstream-timeout
        upstream-unreachable
        upstream-error
      ].freeze

      OBSERVATION_SCHEMA = "flexdoc.host-execution.observation/1"

      # Evidence an API host cannot observe by itself, declared rather than omitted.
      OBSERVATION_GAPS = %w[browser-direct-transport-mix].freeze

      OUTCOMES = %w[success rejected error].freeze

      def self.host_execution_reason?(value)
        HOST_EXECUTION_REASONS.include?(value)
      end
    end

    # One dependency-free metric update that can be bridged to Prometheus or
    # OpenTelemetry without FlexDoc owning a registry.
    HostExecutionMetric = Data.define(:name, :kind, :value, :labels) do
      def initialize(name:, kind:, value:, labels: {})
        super
      end
    end

    # Aggregates host-execution evidence for one window, free of request content.
    #
    # Only counters and durations are retained. No URL, header, body, credential
    # or per-request timestamp reaches this recorder, so the aggregate cannot
    # carry request content by construction.
    #
    # Durations are retained up to +duration_sample_capacity+ and then replaced by
    # reservoir sampling, so memory stays bounded for a host that runs
    # indefinitely while percentiles stay representative of the whole window
    # rather than only its opening. Counts stay exact regardless.
    class HostExecutionObservation
      def initialize(duration_sample_capacity: 8192, clock: -> { Time.now }, random: -> { Kernel.rand })
        @capacity = [duration_sample_capacity.to_i, 1].max
        @clock = clock
        @random = random
        @mutex = Mutex.new
        reset
      end

      # Discard all counts and start a new window.
      def reset
        @mutex.synchronize { clear_state }
        nil
      end

      # Fold one metric update into the aggregate. Safe to pass as a metric sink
      # from concurrent request handlers.
      def record(metric)
        @mutex.synchronize do
          timestamp = @clock.call
          @window_start ||= timestamp
          @window_end = timestamp

          case metric.name
          when "flexdoc_execute_requests_total"
            @started += 1
          when "flexdoc_execute_in_flight"
            @in_flight = [@in_flight + metric.value.to_i, 0].max
            @peak_in_flight = [@peak_in_flight, @in_flight].max
          when "flexdoc_execute_completions_total"
            @completed += 1
            outcome = metric.labels["outcome"] || metric.labels[:outcome]
            @outcomes[outcome] += 1 if @outcomes.key?(outcome)
          when "flexdoc_execute_rejections_total"
            tally(@rejections, metric.labels["reason"] || metric.labels[:reason])
          when "flexdoc_execute_unmarked_total"
            # Deliberately not folded into rejections: those describe validated
            # envelopes, and merging the two would double-count attempts.
            @unmarked += 1
          when "flexdoc_execute_errors_total"
            tally(@errors, metric.labels["reason"] || metric.labels[:reason])
          when "flexdoc_execute_duration_seconds"
            record_duration(metric.value.to_f)
          end
        end
        nil
      end

      # A callable sink usable directly as the executor's +metric_sink+.
      def sink
        method(:record)
      end

      # Current aggregate; safe to call at any time.
      def snapshot
        @mutex.synchronize do
          ordered = @durations.sort
          {
            "windowStart" => @window_start&.utc&.iso8601(3),
            "windowEnd" => @window_end&.utc&.iso8601(3),
            "startedExecutions" => @started,
            "unmarkedRequests" => @unmarked,
            "completedExecutions" => @completed,
            "outcomes" => @outcomes.dup,
            "inFlight" => @in_flight,
            "peakInFlight" => @peak_in_flight,
            "rejectionsByReason" => @rejections.dup,
            "errorsByReason" => @errors.dup,
            "durations" => duration_summary(ordered)
          }
        end
      end

      private

      def clear_state
        @window_start = nil
        @window_end = nil
        @started = 0
        @unmarked = 0
        @completed = 0
        @in_flight = 0
        @peak_in_flight = 0
        @observed_durations = 0
        @outcomes = HostExecutionObservability::OUTCOMES.to_h { |outcome| [outcome, 0] }
        @rejections = {}
        @errors = {}
        @durations = []
      end

      def tally(counts, reason)
        return unless HostExecutionObservability.host_execution_reason?(reason)

        counts[reason] = counts.fetch(reason, 0) + 1
      end

      def record_duration(seconds)
        milliseconds = [seconds * 1000, 0.0].max
        @observed_durations += 1
        if @durations.length < @capacity
          @durations << milliseconds
          return
        end
        candidate = (@random.call * @observed_durations).to_i
        @durations[candidate] = milliseconds if candidate < @capacity
      end

      def duration_summary(ordered)
        return nil if ordered.empty?

        {
          "sampleCount" => ordered.length,
          "sampled" => @observed_durations > ordered.length,
          "minMs" => ordered.first,
          "p50Ms" => percentile(ordered, 0.5),
          "p95Ms" => percentile(ordered, 0.95),
          "p99Ms" => percentile(ordered, 0.99),
          "maxMs" => ordered.last
        }
      end

      def percentile(ordered, fraction)
        return ordered.first if ordered.length == 1

        rank = (fraction * ordered.length).ceil.clamp(1, ordered.length)
        ordered[rank - 1]
      end
    end

    # Build the operator export document for a host-execution observation.
    #
    # The document is aggregate-only and is meant to be written to disk or handed
    # to an operator; FlexDoc never transmits it. Browser-direct executions never
    # reach an API host, so the transport mix cannot be derived here, and that gap
    # is declared so a review cannot mistake this document for complete evidence.
    def self.host_execution_observation_report(observation, generated_at: nil)
      {
        "schema" => HostExecutionObservability::OBSERVATION_SCHEMA,
        "generatedAt" => generated_at || Time.now.utc.iso8601(3),
        "runtime" => "ruby",
        "observation" => observation.snapshot,
        "gaps" => HostExecutionObservability::OBSERVATION_GAPS.dup
      }
    end
  end
end

File: adapters/go/host_execution_observability.go
package flexdoc

import (
	"math"
	"math/rand"
	"sort"
	"sync"
	"time"
)

// This mirrors the Node and Python contract deliberately: the same reason
// vocabulary, the same metric names and labels, and the same export schema. An
// operator running a mixed fleet should read one document shape regardless of
// which runtime served the execute route, and a collector written for one
// runtime should not need a second parser for another.

// HostExecutionReason is a stable low-cardinality category for a non-successful execution.
//
// Rejection messages interpolate request values such as origins, field names
// and methods, so they are unbounded and cannot be used as a metric label.
// These can.
type HostExecutionReason string

// Stable reason categories shared with every other FlexDoc runtime.
const (
	ReasonMarkerMissing        HostExecutionReason = "marker-missing"
	ReasonExecutionDisabled    HostExecutionReason = "execution-disabled"
	ReasonAdmissionSaturated   HostExecutionReason = "admission-saturated"
	ReasonDestinationForbidden HostExecutionReason = "destination-forbidden"
	ReasonRedirectForbidden    HostExecutionReason = "redirect-forbidden"
	ReasonBodyMalformed        HostExecutionReason = "body-malformed"
	ReasonBodyTooLarge         HostExecutionReason = "body-too-large"
	ReasonUnsupportedMediaType HostExecutionReason = "unsupported-media-type"
	ReasonRequestInvalid       HostExecutionReason = "request-invalid"
	ReasonAuthUnsupported      HostExecutionReason = "auth-unsupported"
	ReasonUpstreamTimeout      HostExecutionReason = "upstream-timeout"
	ReasonUpstreamUnreachable  HostExecutionReason = "upstream-unreachable"
	ReasonUpstreamError        HostExecutionReason = "upstream-error"
)

// ObservationSchema identifies the operator export document shared across runtimes.
const ObservationSchema = "flexdoc.host-execution.observation/1"

var hostExecutionReasons = []HostExecutionReason{
	ReasonMarkerMissing, ReasonExecutionDisabled, ReasonAdmissionSaturated,
	ReasonDestinationForbidden, ReasonRedirectForbidden, ReasonBodyMalformed,
	ReasonBodyTooLarge, ReasonUnsupportedMediaType, ReasonRequestInvalid,
	ReasonAuthUnsupported, ReasonUpstreamTimeout, ReasonUpstreamUnreachable,
	ReasonUpstreamError,
}

// observationGaps names evidence an API host cannot observe by itself.
var observationGaps = []string{"browser-direct-transport-mix"}

// HostExecutionReasons returns the stable reason categories in contract order.
func HostExecutionReasons() []HostExecutionReason {
	out := make([]HostExecutionReason, len(hostExecutionReasons))
	copy(out, hostExecutionReasons)
	return out
}

// IsHostExecutionReason reports whether a value is one of the stable categories.
func IsHostExecutionReason(value string) bool {
	for _, reason := range hostExecutionReasons {
		if string(reason) == value {
			return true
		}
	}
	return false
}

// HostExecutionMetric is one dependency-free metric update that can be bridged
// to Prometheus or OpenTelemetry without FlexDoc owning a registry.
type HostExecutionMetric struct {
	Name   string
	Kind   string
	Value  float64
	Labels map[string]string
}

// MetricSink receives metric updates. Delivery is best effort: a sink that
// panics cannot fail an execution.
type MetricSink func(HostExecutionMetric)

// HostExecutionDurations is the duration distribution for one observation window.
type HostExecutionDurations struct {
	SampleCount int     `json:"sampleCount"`
	Sampled     bool    `json:"sampled"`
	MinMs       float64 `json:"minMs"`
	P50Ms       float64 `json:"p50Ms"`
	P95Ms       float64 `json:"p95Ms"`
	P99Ms       float64 `json:"p99Ms"`
	MaxMs       float64 `json:"maxMs"`
}

// HostExecutionSnapshot is the aggregate for one observation window.
type HostExecutionSnapshot struct {
	WindowStart         *string                 `json:"windowStart"`
	WindowEnd           *string                 `json:"windowEnd"`
	StartedExecutions   int                     `json:"startedExecutions"`
	UnmarkedRequests    int                     `json:"unmarkedRequests"`
	CompletedExecutions int                     `json:"completedExecutions"`
	Outcomes            map[string]int          `json:"outcomes"`
	InFlight            int                     `json:"inFlight"`
	PeakInFlight        int                     `json:"peakInFlight"`
	RejectionsByReason  map[string]int          `json:"rejectionsByReason"`
	ErrorsByReason      map[string]int          `json:"errorsByReason"`
	Durations           *HostExecutionDurations `json:"durations"`
}

// HostExecutionObservationReport is the operator export document.
type HostExecutionObservationReport struct {
	Schema      string                `json:"schema"`
	GeneratedAt string                `json:"generatedAt"`
	Runtime     string                `json:"runtime"`
	Observation HostExecutionSnapshot `json:"observation"`
	Gaps        []string              `json:"gaps"`
}

// HostExecutionObservation aggregates execution evidence for one window.
//
// Only counters and durations are retained. No URL, header, body, credential or
// per-request timestamp reaches this recorder, so the aggregate cannot carry
// request content by construction.
//
// Durations are retained up to the sample capacity and then replaced by
// reservoir sampling, so memory stays bounded for a host that runs indefinitely
// while percentiles stay representative of the whole window rather than only
// its opening. Counts stay exact regardless.
type HostExecutionObservation struct {
	mu                sync.Mutex
	capacity          int
	now               func() time.Time
	random            func() float64
	windowStart       *time.Time
	windowEnd         *time.Time
	started           int
	unmarked          int
	completed         int
	inFlight          int
	peakInFlight      int
	observedDurations int
	outcomes          map[string]int
	rejections        map[string]int
	errors            map[string]int
	durations         []float64
}

// NewHostExecutionObservation creates a recorder with the default 8192-sample capacity.
func NewHostExecutionObservation() *HostExecutionObservation {
	return NewHostExecutionObservationWithCapacity(8192)
}

// NewHostExecutionObservationWithCapacity creates a recorder retaining at most
// capacity durations before uniform sampling begins.
func NewHostExecutionObservationWithCapacity(capacity int) *HostExecutionObservation {
	if capacity < 1 {
		capacity = 1
	}
	observation := &HostExecutionObservation{
		capacity: capacity,
		now:      time.Now,
		random:   rand.Float64,
	}
	observation.Reset()
	return observation
}

// Reset discards all counts and starts a new window.
func (o *HostExecutionObservation) Reset() {
	o.mu.Lock()
	defer o.mu.Unlock()
	o.windowStart = nil
	o.windowEnd = nil
	o.started = 0
	o.unmarked = 0
	o.completed = 0
	o.inFlight = 0
	o.peakInFlight = 0
	o.observedDurations = 0
	o.outcomes = map[string]int{"success": 0, "rejected": 0, "error": 0}
	o.rejections = map[string]int{}
	o.errors = map[string]int{}
	o.durations = nil
}

// Record folds one metric update into the aggregate. It is safe to use as a
// MetricSink from concurrent request handlers.
func (o *HostExecutionObservation) Record(metric HostExecutionMetric) {
	o.mu.Lock()
	defer o.mu.Unlock()

	timestamp := o.now()
	if o.windowStart == nil {
		start := timestamp
		o.windowStart = &start
	}
	end := timestamp
	o.windowEnd = &end

	switch metric.Name {
	case "flexdoc_execute_requests_total":
		o.started++
	case "flexdoc_execute_in_flight":
		o.inFlight += int(metric.Value)
		if o.inFlight < 0 {
			o.inFlight = 0
		}
		if o.inFlight > o.peakInFlight {
			o.peakInFlight = o.inFlight
		}
	case "flexdoc_execute_completions_total":
		o.completed++
		if _, known := o.outcomes[metric.Labels["outcome"]]; known {
			o.outcomes[metric.Labels["outcome"]]++
		}
	case "flexdoc_execute_rejections_total":
		o.tally(o.rejections, metric.Labels["reason"])
	case "flexdoc_execute_unmarked_total":
		// Deliberately not folded into rejections: those describe validated
		// envelopes, and merging the two would double-count attempts.
		o.unmarked++
	case "flexdoc_execute_errors_total":
		o.tally(o.errors, metric.Labels["reason"])
	case "flexdoc_execute_duration_seconds":
		o.recordDuration(metric.Value)
	}
}

// Snapshot returns the current aggregate; safe to call at any time.
func (o *HostExecutionObservation) Snapshot() HostExecutionSnapshot {
	o.mu.Lock()
	defer o.mu.Unlock()

	snapshot := HostExecutionSnapshot{
		WindowStart:         formatInstant(o.windowStart),
		WindowEnd:           formatInstant(o.windowEnd),
		StartedExecutions:   o.started,
		UnmarkedRequests:    o.unmarked,
		CompletedExecutions: o.completed,
		Outcomes:            copyCounts(o.outcomes),
		InFlight:            o.inFlight,
		PeakInFlight:        o.peakInFlight,
		RejectionsByReason:  copyCounts(o.rejections),
		ErrorsByReason:      copyCounts(o.errors),
	}
	if len(o.durations) == 0 {
		return snapshot
	}

	ordered := make([]float64, len(o.durations))
	copy(ordered, o.durations)
	sort.Float64s(ordered)
	snapshot.Durations = &HostExecutionDurations{
		SampleCount: len(ordered),
		Sampled:     o.observedDurations > len(ordered),
		MinMs:       ordered[0],
		P50Ms:       percentile(ordered, 0.5),
		P95Ms:       percentile(ordered, 0.95),
		P99Ms:       percentile(ordered, 0.99),
		MaxMs:       ordered[len(ordered)-1],
	}
	return snapshot
}

// NewHostExecutionObservationReport builds the operator export document.
//
// The document is aggregate-only and is meant to be written to disk or handed
// to an operator; FlexDoc never transmits it. Browser-direct executions never
// reach an API host, so the transport mix cannot be derived here, and that gap
// is declared so a review cannot mistake this document for complete evidence.
func NewHostExecutionObservationReport(observation *HostExecutionObservation) HostExecutionObservationReport {
	gaps := make([]string, len(observationGaps))
	copy(gaps, observationGaps)
	return HostExecutionObservationReport{
		Schema:      ObservationSchema,
		GeneratedAt: time.Now().UTC().Format(time.RFC3339Nano),
		Runtime:     "go",
		Observation: observation.Snapshot(),
		Gaps:        gaps,
	}
}

func (o *HostExecutionObservation) tally(counts map[string]int, reason string) {
	if IsHostExecutionReason(reason) {
		counts[reason]++
	}
}

func (o *HostExecutionObservation) recordDuration(seconds float64) {
	milliseconds := math.Max(0, seconds*1000)
	o.observedDurations++
	if len(o.durations) < o.capacity {
		o.durations = append(o.durations, milliseconds)
		return
	}
	candidate := int(o.random() * float64(o.observedDurations))
	if candidate < o.capacity {
		o.durations[candidate] = milliseconds
	}
}

func copyCounts(counts map[string]int) map[string]int {
	out := make(map[string]int, len(counts))
	for key, value := range counts {
		out[key] = value
	}
	return out
}

func formatInstant(instant *time.Time) *string {
	if instant == nil {
		return nil
	}
	formatted := instant.UTC().Format(time.RFC3339Nano)
	return &formatted
}

func percentile(sorted []float64, fraction float64) float64 {
	if len(sorted) == 1 {
		return sorted[0]
	}
	rank := int(math.Ceil(fraction * float64(len(sorted))))
	if rank < 1 {
		rank = 1
	}
	if rank > len(sorted) {
		rank = len(sorted)
	}
	return sorted[rank-1]
}
